lib/ex_unit/lib/ex_unit/callbacks.ex

```diff
@@ -373,24 +398,30 @@ defmodule ExUnit.Callbacks do
   end
 
   @doc """
-  Defines a callback that runs once the test exits.
+  Registers a callback that runs once the test exits.
 
   `callback` is a function that receives no arguments and
-  runs in a separate process than the caller.
+  runs in a separate process than the caller. Its return
+  value is irrelevant and is discarded.
 
-  `on_exit/2` is usually called from `setup` and `setup_all`
+  `on_exit/2` is usually called from `setup/1` and `setup_all/1`
   callbacks, often to undo the action performed during the setup.
-  However, `on_exit/2` may also be called dynamically, where a
-  reference can be used to guarantee the callback will be invoked
-  only once.
-
-  `on_exit/2` gets executed in a blocking fashion after a test
-  exits and **before** running the next test. This means that no
-  other test from the same test case will be running while the
-  `on_exit/2` callback for a previous test is running.
 
-  `on_exit/2` is executed in a different process than the test
-  process.
+  However, `on_exit/2` may also be called dynamically. An "ID" (the
+  `name_or_ref` argument) can be used to guarantee that the callback
+  will be invoked only once. ExUnit uses this term to identify a
+  `on_exit/2` handler: if you want to override a previous handler, for
+  example, use the same `name_or_ref` across multiple `on_exit/2`
+  calls.
+
+  If `on_exit/2` is called inside `setup/1` or inside a test, it's
+  executed in a blocking fashion after the test exits and *before
+  running the next test*. This means that no other test from the same
+  test case will be running while the `on_exit/2` callback for a
+  previous test is running. `on_exit/2` is executed in a different
+  process than the test process. On the other hand, if `on_exit/2` is
+  called inside a `setup_all/1` callback then `callback` is executed
+  after running *all tests* (see `setup_all/1` for more information).
```

Comment:
What I understand from this is that even if running with `async: true` if a test finishes, it will enter the blocking state for all the tests within the test case due to `on_exit/2` callback, and:
1. even if other tests are done they will not start until this call back is finished.
2. all other test running concurrently will suspspend _"This means that no other test from the same test case will be running while the `on_exit/2` callback"_

I don't think the 2nd one is what it means, and for the first one i'm not sure.

Reply:
@eksperimental tests in a single test case never run concurrently. Maybe that answers your question, otherwise I'm not sure what you mean or how we could improve this documentation?